wandb/sdk/lib/run_moment.py

```diff
@@ -0,0 +1,74 @@
+from dataclasses import dataclass
+from typing import Literal, Union, cast
+from urllib import parse
+
+_STEP = Literal["_step"]
+
+
+@dataclass
+class RunMoment:
+    """A moment in a run."""
+
+    run: str  # run name
+
+    # currently, the _step value to fork from. in future, this will be optional
+    value: Union[int, float]
+
+    # only step for now, in future this will be relaxed to be any metric
+    metric: _STEP = "_step"
+
+    def __post_init__(self):
+        if self.metric != "_step":
+            raise ValueError(
+                f"Only the metric '_step' is supported, got '{self.metric}'."
+            )
+        if not isinstance(self.value, (int, float)):
+            raise ValueError(
+                f"Only int or float values are supported, got '{self.value}'."
+            )
+        if not isinstance(self.run, str):
+            raise ValueError(f"Only string run names are supported, got '{self.run}'.")
+
+    @classmethod
+    def from_uri(cls, uri: str) -> "RunMoment":
+        parsable = "runmoment://" + uri
+        parse_err = ValueError(
+            f"Could not parse passed run moment string '{uri}', "
+            f"expected format '<run>?<metric>=<numeric_value>'. "
+            f"Currently, only the metric '_step' is supported. "
+            f"Example: 'ans3bsax?_step=123'."
+        )
+
+        try:
+            parsed = parse.urlparse(parsable)
```

Comment:
This adds a lot of irrelevant URL complexity, like schemas and fragments. The format is simple enough for a regex:

```py
match = re.match(r"(.+)\?_step=(.+)", value)
if not match:
    raise parse_err

run_id = match.group(1)
step_str = match.group(2)

try:
    step = int(step_str) if step_str.isdigit() else float(step_str)
except ValueError as e:
    raise parse_err from e

return cls(
    run=run_id,
    metric="_step",
    value=step,
)
```

Reply:
this doesn't work, these tests fail:

```
========================================================================================================= FAILURES =========================================================================================================
__________________________________________________________________________________________ test_run_moment_from_uri_invalid_path ___________________________________________________________________________________________

    def test_run_moment_from_uri_invalid_path():
        uri = "ans3bsax/metric?_step=123"
>       with pytest.raises(ValueError):
E       Failed: DID NOT RAISE <class 'ValueError'>

/Users/danielgoldstein/go/src/github.com/wandb/wandb/tests/pytest_tests/unit_tests/test_run_moment.py:33: Failed
_________________________________________________________________________________________ test_run_moment_from_uri_invalid_scheme __________________________________________________________________________________________

    def test_run_moment_from_uri_invalid_scheme():
        uri = "http://ans3bsax?_step=123"
>       with pytest.raises(ValueError):
E       Failed: DID NOT RAISE <class 'ValueError'>

/Users/danielgoldstein/go/src/github.com/wandb/wandb/tests/pytest_tests/unit_tests/test_run_moment.py:51: Failed
================================================================================================= short test summary info ==================================================================================================
FAILED tests/pytest_tests/unit_tests/test_run_moment.py::test_run_moment_from_uri_invalid_path - Failed: DID NOT RAISE <class 'ValueError'>
FAILED tests/pytest_tests/unit_tests/test_run_moment.py::test_run_moment_from_uri_invalid_scheme - Failed: DID NOT RAISE <class 'ValueError'>
```

the second one is a bit of a contrived case but the first is realistic. since we want these to obey URI parsing behavior with respect to special characters, slashes, etc., I think we should just keep this as is

Reply:
these tests can be made to pass by using a smarter regex:

```py
# Disallow the special characters mentioned in the documentation:
run_re = r"[^/\#?%:]+"
match = re.match(fr"({run_re})\?_step=(\d+)", value)
```

I guess we also expect step to be an integer only, right? So I updated its regex to be `\d+` rather than `.+`.

Reply:
this will work for now but will break once we allow forking from runs in other projects. e.g., 

fork from 

```
e1/p1/r1?_metric=20.0
```

the regex will need to be updated to allow e1, p1, r1 to all get through and we will need to extract them from the resulting group. its all doable but it seems like we are just going down the path of re implementing url parsing rules with regex. i'd rather use a well tested builtin implementation of a url parser than roll my own using regex

Reply:
> Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems.

Reply:
It's okay to update code. It doesn't seem that we actually want this to be a URL; rather it's just a URL-like format. That forces you to add an assertion that there's no fragment, and that there's only one query parameter. For simple formats, a regular expression can be the right tool; if it were any simpler you could even use `uri.split(...)`.

If you do keep it this way, could you reduce the nesting in the last block?

```py
if len(query) != 1:
    raise parse_err

# No else needed.
metric = list(query.keys())[0]
...

# Shorten the isdigit() conditional
try:
    num_value = int(value) if value.isdigit() else float(value)
except ValueError as e:
    ....

```

Reply:
done